### ISSUE TITLE
cuSignal 0.15 Performance Comparison

### ISSUE THREAD

**Issue by mnicely** [rapidsai/cusignal#183]
Display performance number on V100 comparing cuSignal 0.14 and cusignal 0.15, using pytest benchmarks.

**Comment by mnicely** [comment]
## Comparison
cuSignal 0.15 is tagged with *0001_0.15_cu*
cuSignal 0.14 is tagged with *0002_0.14_cu*

Due to a change in 0.15, tests are now prefixed with _test_ versus _bench_, in 0.14.

Also, 0.15 doesn't use Numba any longer. Therefore, in 0.14 result **False** means **no Numba**.

### Doc
[cusignal_0-14_0-15_comparison.txt](https://github.com/rapidsai/cusignal/files/5085733/cusignal_0-14_0-15_comparison.txt)

### Example
```bash
--------------------------------------------------------------------------------------------------- benchmark 'UpFirDn': 54 tests ---------------------------------------------------------------------------------------------------
Name (time in us)                                             Min                   Max                Mean             StdDev              Median                IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_upfirdn_gpu[-1-1-3-16384] (0001_0.15_cu)            316.4271 (1.0)        497.0962 (1.28)     325.7232 (1.00)      7.7458 (1.68)     324.3871 (1.0)       2.4848 (1.08)      194;320        3.0701 (1.00)       3038           1
test_upfirdn_gpu[-1-1-7-16384] (0001_0.15_cu)            317.0129 (1.00)       564.7978 (1.45)     328.2977 (1.01)     12.3999 (2.69)     325.2001 (1.00)      2.6962 (1.17)      180;437        3.0460 (0.99)       3063           1
test_upfirdn_gpu[-1-2-3-16384] (0001_0.15_cu)            317.3575 (1.00)       424.3981 (1.09)     326.7542 (1.00)      7.1077 (1.54)     325.1592 (1.00)      2.5043 (1.09)      227;334        3.0604 (1.00)       3051           1
test_upfirdn_gpu[-1-2-7-16384] (0001_0.15_cu)            317.8269 (1.00)       675.0179 (1.74)     351.5181 (1.08)     34.8741 (7.57)     336.8291 (1.04)     43.5077 (18.88)      457;77        2.8448 (0.93)       3034           1
test_upfirdn_gpu[0-2-7-16384] (0001_0.15_cu)             318.3335 (1.01)       536.1997 (1.38)     333.6300 (1.02)     23.0654 (5.01)     326.3131 (1.01)      3.4566 (1.50)      252;569        2.9973 (0.98)       3037           1
test_upfirdn_gpu[0-9-7-16384] (0001_0.15_cu)             319.4744 (1.01)       426.6379 (1.10)     327.3932 (1.01)      5.7443 (1.25)     326.0858 (1.01)      2.4324 (1.06)      311;345        3.0544 (0.99)       3049           1
test_upfirdn_gpu[0-1-3-16384] (0001_0.15_cu)             319.5629 (1.01)       506.1533 (1.30)     332.9212 (1.02)     20.6603 (4.49)     326.0281 (1.01)      4.0659 (1.76)      251;586        3.0037 (0.98)       3037           1
test_upfirdn_gpu[-1-9-3-16384] (0001_0.15_cu)            319.5796 (1.01)       388.9436 (1.0)      325.6150 (1.0)       4.6065 (1.0)      324.6050 (1.00)      2.5025 (1.09)      343;317        3.0711 (1.0)        3059           1
test_upfirdn_gpu[0-9-3-16384] (0001_0.15_cu)             320.2017 (1.01)       516.9213 (1.33)     331.5146 (1.02)     19.9277 (4.33)     325.6104 (1.00)      3.2969 (1.43)      215;532        3.0165 (0.98)       3046           1
test_upfirdn_gpu[-1-9-7-16384] (0001_0.15_cu)            320.2241 (1.01)       509.5704 (1.31)     331.2279 (1.02)     19.1754 (4.16)     325.8074 (1.00)      2.9714 (1.29)      198;506        3.0191 (0.98)       3061           1
test_upfirdn_gpu[0-1-7-16384] (0001_0.15_cu)             320.5789 (1.01)       483.0239 (1.24)     327.6336 (1.01)      6.2080 (1.35)     326.3960 (1.01)      2.6918 (1.17)      265;303        3.0522 (0.99)       3045           1
test_upfirdn_gpu[0-2-3-16384] (0001_0.15_cu)             320.8155 (1.01)       397.3506 (1.02)     328.0229 (1.01)      6.2529 (1.36)     326.4965 (1.01)      2.6361 (1.14)      300;336        3.0486 (0.99)       3040           1
bench_upfirdn_gpu[False--1-9-7-16384] (0002_0.14_cu)     333.7050 (1.05)     2,648.1124 (6.81)     350.0518 (1.08)     48.7688 (10.59)    341.9621 (1.05)      3.0759 (1.33)      156;527        2.8567 (0.93)       2901           1
bench_upfirdn_gpu[False-0-1-3-16384] (0002_0.14_cu)      334.5823 (1.06)       623.4990 (1.60)     349.7612 (1.07)     22.4921 (4.88)     342.8655 (1.06)      2.9269 (1.27)      246;518        2.8591 (0.93)       2893           1
bench_upfirdn_gpu[False-0-9-3-16384] (0002_0.14_cu)      335.0107 (1.06)       540.7510 (1.39)     348.4780 (1.07)     20.3236 (4.41)     342.6373 (1.06)      2.7637 (1.20)      207;481        2.8696 (0.93)       2896           1
bench_upfirdn_gpu[False-0-9-7-16384] (0002_0.14_cu)      335.0172 (1.06)       532.3403 (1.37)     346.7570 (1.06)     15.3340 (3.33)     342.5162 (1.06)      2.8852 (1.25)      187;493        2.8839 (0.94)       2904           1
bench_upfirdn_gpu[False--1-1-3-16384] (0002_0.14_cu)     335.2519 (1.06)       548.8023 (1.41)     343.7147 (1.06)      9.2314 (2.00)     341.8494 (1.05)      2.3046 (1.0)       110;326        2.9094 (0.95)       2875           1
bench_upfirdn_gpu[False-0-2-3-16384] (0002_0.14_cu)      335.3450 (1.06)       505.4902 (1.30)     346.8858 (1.07)     15.6832 (3.40)     342.5255 (1.06)      2.8107 (1.22)      182;485        2.8828 (0.94)       2902           1
bench_upfirdn_gpu[False--1-1-7-16384] (0002_0.14_cu)     335.6552 (1.06)       623.0269 (1.60)     347.2372 (1.07)     17.2092 (3.74)     342.9502 (1.06)      2.7220 (1.18)      158;443        2.8799 (0.94)       2893           1
bench_upfirdn_gpu[False-0-2-7-16384] (0002_0.14_cu)      336.0817 (1.06)       694.8924 (1.79)     352.5666 (1.08)     30.7355 (6.67)     342.4422 (1.06)      4.3465 (1.89)      212;441        2.8363 (0.92)       2472           1
bench_upfirdn_gpu[False--1-2-7-16384] (0002_0.14_cu)     337.4182 (1.07)       530.5512 (1.36)     343.2602 (1.05)      7.9440 (1.72)     341.5616 (1.05)      2.4689 (1.07)      208;351        2.9132 (0.95)       2890           1
bench_upfirdn_gpu[False--1-2-3-16384] (0002_0.14_cu)     337.4480 (1.07)       536.2732 (1.38)     344.4452 (1.06)     11.8968 (2.58)     341.3958 (1.05)      2.4789 (1.08)      160;423        2.9032 (0.95)       2889           1
bench_upfirdn_gpu[False--1-9-3-16384] (0002_0.14_cu)     337.6547 (1.07)       475.4635 (1.22)     343.9021 (1.06)      8.1530 (1.77)     342.0608 (1.05)      2.3241 (1.01)      191;370        2.9078 (0.95)       2896           1
bench_upfirdn_gpu[False-0-1-7-16384] (0002_0.14_cu)      338.4640 (1.07)       638.1199 (1.64)     346.5221 (1.06)     15.7915 (3.43)     342.8832 (1.06)      2.6901 (1.17)      145;420        2.8858 (0.94)       2897           1
test_upfirdn_gpu[0-1-2-16384] (0001_0.15_cu)             346.1586 (1.09)       539.0523 (1.39)     357.1259 (1.10)      6.9925 (1.52)     355.6469 (1.10)      2.7409 (1.19)      273;334        2.8001 (0.91)       2782           1
test_upfirdn_gpu[-1-1-2-16384] (0001_0.15_cu)            346.8227 (1.10)     2,921.6260 (7.51)     357.9073 (1.10)     49.7099 (10.79)    354.8097 (1.09)      2.8394 (1.23)       19;388        2.7940 (0.91)       2777           1
test_upfirdn_gpu[-1-2-2-16384] (0001_0.15_cu)            349.1156 (1.10)       518.7215 (1.33)     356.4598 (1.09)      6.8313 (1.48)     354.9866 (1.09)      2.4999 (1.08)      258;311        2.8054 (0.91)       2789           1
test_upfirdn_gpu[-1-9-2-16384] (0001_0.15_cu)            349.7479 (1.11)       570.8523 (1.47)     362.1593 (1.11)     22.0988 (4.80)     355.3974 (1.10)      3.3251 (1.44)      220;499        2.7612 (0.90)       2803           1
test_upfirdn_gpu[0-9-2-16384] (0001_0.15_cu)             350.2611 (1.11)     2,657.9555 (6.83)     357.9042 (1.10)     44.5402 (9.67)     355.3936 (1.10)      2.3888 (1.04)       14;318        2.7940 (0.91)       2783           1
test_upfirdn_gpu[0-2-2-16384] (0001_0.15_cu)             351.0658 (1.11)       533.0956 (1.37)     362.9515 (1.11)     22.9432 (4.98)     356.2588 (1.10)      3.3807 (1.47)      195;463        2.7552 (0.90)       2778           1
bench_upfirdn_gpu[False--1-9-2-16384] (0002_0.14_cu)     364.1071 (1.15)       576.9581 (1.48)     384.4172 (1.18)     27.1936 (5.90)     373.9232 (1.15)      4.6650 (2.02)      299;466        2.6013 (0.85)       2637           1
bench_upfirdn_gpu[False-0-1-2-16384] (0002_0.14_cu)      365.3625 (1.15)       539.3587 (1.39)     374.7798 (1.15)      9.1545 (1.99)     372.8671 (1.15)      2.6107 (1.13)      119;324        2.6682 (0.87)       2649           1
bench_upfirdn_gpu[False--1-1-2-16384] (0002_0.14_cu)     366.4857 (1.16)       578.7835 (1.49)     384.0599 (1.18)     25.9696 (5.64)     374.6441 (1.15)      4.8075 (2.09)      270;455        2.6038 (0.85)       2644           1
bench_upfirdn_gpu[False--1-2-2-16384] (0002_0.14_cu)     367.7197 (1.16)       578.7937 (1.49)     378.3894 (1.16)     14.9368 (3.24)     373.0226 (1.15)      3.5674 (1.55)      273;530        2.6428 (0.86)       2657           1
bench_upfirdn_gpu[False-0-9-2-16384] (0002_0.14_cu)      367.7504 (1.16)     2,730.0594 (7.02)     380.3794 (1.17)     48.9108 (10.62)    373.8618 (1.15)      3.0822 (1.34)      149;444        2.6290 (0.86)       2669           1
bench_upfirdn_gpu[False-0-2-2-16384] (0002_0.14_cu)      369.4920 (1.17)       573.7348 (1.48)     375.5821 (1.15)      8.7774 (1.91)     373.6839 (1.15)      2.3174 (1.01)      124;336        2.6625 (0.87)       2651           1
bench_upfirdn_gpu[True--1-9-7-16384] (0002_0.14_cu)      682.2189 (2.16)       910.5681 (2.34)     701.9639 (2.16)     27.7599 (6.03)     692.4421 (2.13)      9.1991 (3.99)      137;180        1.4246 (0.46)       1451           1
bench_upfirdn_gpu[True--1-1-7-16384] (0002_0.14_cu)      683.4371 (2.16)     2,099.4032 (5.40)     697.4666 (2.14)     39.4512 (8.56)     693.1107 (2.14)      4.2561 (1.85)       32;198        1.4338 (0.47)       1448           1
bench_upfirdn_gpu[True-0-1-3-16384] (0002_0.14_cu)       683.7221 (2.16)     3,003.6727 (7.72)     709.4437 (2.18)     71.6390 (15.55)    694.6465 (2.14)     10.9943 (4.77)       76;175        1.4096 (0.46)       1449           1
bench_upfirdn_gpu[True-0-2-3-16384] (0002_0.14_cu)       684.4606 (2.16)     1,114.7680 (2.87)     714.8540 (2.20)     73.4121 (15.94)    692.5976 (2.14)      8.8359 (3.83)       95;196        1.3989 (0.46)       1447           1
bench_upfirdn_gpu[True--1-2-3-16384] (0002_0.14_cu)      684.5882 (2.16)       906.1899 (2.33)     698.7973 (2.15)     23.7471 (5.16)     691.5382 (2.13)      7.7966 (3.38)      100;146        1.4310 (0.47)       1448           1
bench_upfirdn_gpu[True-0-9-3-16384] (0002_0.14_cu)       684.7633 (2.16)       931.0003 (2.39)     705.2597 (2.17)     35.7589 (7.76)     692.7899 (2.14)     10.6730 (4.63)      118;186        1.4179 (0.46)       1450           1
bench_upfirdn_gpu[True--1-9-3-16384] (0002_0.14_cu)      685.4944 (2.17)       888.5646 (2.28)     695.7371 (2.14)     11.3797 (2.47)     692.9282 (2.14)      4.1618 (1.81)       56;214        1.4373 (0.47)       1441           1
bench_upfirdn_gpu[True--1-2-7-16384] (0002_0.14_cu)      685.6471 (2.17)     1,101.0254 (2.83)     702.5043 (2.16)     36.2521 (7.87)     692.5347 (2.13)      7.3882 (3.21)       88;141        1.4235 (0.46)       1434           1
bench_upfirdn_gpu[True-0-1-7-16384] (0002_0.14_cu)       685.7496 (2.17)       888.0859 (2.28)     696.1332 (2.14)     11.4490 (2.49)     693.4628 (2.14)      4.0072 (1.74)       47;220        1.4365 (0.47)       1449           1
bench_upfirdn_gpu[True-0-9-7-16384] (0002_0.14_cu)       686.2739 (2.17)     1,111.0753 (2.86)     741.1770 (2.28)     62.1666 (13.50)    703.9560 (2.17)     92.9339 (40.33)      272;14        1.3492 (0.44)       1446           1
bench_upfirdn_gpu[True--1-1-3-16384] (0002_0.14_cu)      686.3158 (2.17)       962.4856 (2.47)     708.9237 (2.18)     41.7359 (9.06)     693.4190 (2.14)     12.3959 (5.38)      125;185        1.4106 (0.46)       1450           1
bench_upfirdn_gpu[True-0-2-7-16384] (0002_0.14_cu)       688.8844 (2.18)     2,822.5007 (7.26)     710.2484 (2.18)     63.7268 (13.83)    696.7513 (2.15)     17.0060 (7.38)       49;132        1.4080 (0.46)       1450           1
bench_upfirdn_gpu[True--1-2-2-16384] (0002_0.14_cu)      715.3293 (2.26)     1,157.4784 (2.98)     776.1658 (2.38)     69.6589 (15.12)    736.9407 (2.27)     88.9548 (38.60)      203;32        1.2884 (0.42)       1384           1
bench_upfirdn_gpu[True-0-1-2-16384] (0002_0.14_cu)       715.5584 (2.26)     1,014.0808 (2.61)     738.6113 (2.27)     25.2194 (5.47)     728.5667 (2.25)     18.6646 (8.10)       102;83        1.3539 (0.44)       1380           1
bench_upfirdn_gpu[True--1-9-2-16384] (0002_0.14_cu)      716.8725 (2.27)     2,858.2728 (7.35)     739.3371 (2.27)     71.0174 (15.42)    725.6325 (2.24)      8.9647 (3.89)       66;148        1.3526 (0.44)       1381           1
bench_upfirdn_gpu[True-0-2-2-16384] (0002_0.14_cu)       718.8777 (2.27)     1,017.8108 (2.62)     740.4537 (2.27)     33.1840 (7.20)     727.6153 (2.24)     12.1023 (5.25)      139;183        1.3505 (0.44)       1377           1
bench_upfirdn_gpu[True--1-1-2-16384] (0002_0.14_cu)      718.9158 (2.27)     1,397.0574 (3.59)     733.0232 (2.25)     23.6017 (5.12)     728.8479 (2.25)      4.7199 (2.05)       42;185        1.3642 (0.44)       1374           1
bench_upfirdn_gpu[True-0-9-2-16384] (0002_0.14_cu)       721.2944 (2.28)       930.0783 (2.39)     731.2569 (2.25)     14.0730 (3.06)     728.0130 (2.24)      4.6808 (2.03)       54;179        1.3675 (0.45)       1378           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

**Comment by awthomp** [comment]
Hey @mnicely -- could you provide a bit more guidance on what we're looking at here? :)

**Comment by datametrician** [comment]
Also could you elaborate on how you no longer use numba?

**Comment by awthomp** [comment]
@datametrician -- Numba is still used, especially to access shared memory. We still highly encourage its use for custom CUDA kernels as well. However, we realized that we could get a sizable speedup over Numba CUDA with raw CuPy CUDA modules. Here, we're basically writing raw CUDA code and wrapping it in Python. We can also make use of CuPy's template functionality and ability to cache and load pre-compiled kernels; we also hope to be able to have these pre-compiled kernels install-able with the library.

Long story short: Numba CUDA for custom kernels and great perf; raw CuPy CUDA for even more perf.

We're not deprecating Numba, but we have deprecated our custom Numba CUDA kernels.

**Comment by mnicely** [comment]
@awthomp I edited my original comment above with more details. Let me know if there's still more information that needs to be added.

@datametrician We discovered that while Numba was a fantastic way to get off the ground, it wasn't as optimal as a CuPy Raw Kernel (CUDA C++). Especially, if data types weren't double complex. I'm presenting at GTC Fall on the techniques we used to go from SciPy -> Numba -> CuPy.

**Comment by datametrician** [comment]
This makes sense. @awthomp I think this will change how we market cuSignal though. Initially, it was a non-compiled library (just CuPy + Numba), but now it seems for max perf we're going deeper into CUDA (at the cost of more engineering time). Is that a far statement?

**Comment by awthomp** [comment]
> This makes sense. @awthomp I think this will change how we market cuSignal though. Initially, it was a non-compiled library (just CuPy + Numba), but now it seems for max perf we're going deeper into CUDA (at the cost of more engineering time). Is that a far statement?

It is! But I still go back to my talk track at SciPy '20:
![image](https://user-images.githubusercontent.com/34922235/90656244-4c8ec300-e210-11ea-8b03-733cb46f22ae.png)

For developers looking to build GPU apps in Python, we recommend drop in replacements. For more perf and customization: Numba. For extreme speed, when needed: raw CUDA. They all build on each other!

Also: cuSignal still doesn't have a dedicated Cython layer, so all our work is still technically pure Python. Admittedly, however, the custom raw CUDA kernels via CuPy are just strings of C++/CUDA code.
